File listing via RPC (to support KBFS mobile integration)

gabriel opened the pull request:
- This moves `kbfs/kbfspath`.go to `libkbfs/path.go` and makes some of it methods public
- Refactors `kbfs/` package to use this
- Creates `libkbfs/fs_rpc.go` which is an RPC implementation of file listing (needed for KBFS integration on mobile)
- Adds `DefaultInitParams` to generate `InitParams` without command line flags (needed for KBFS integration on mobile)

Client changes are at: https://github.com/keybase/client/pull/3179

@akalin-keybase 
cc @strib

gabriel commented:
This won't build until client changes go in.

akalin-keybase commented:
This looks mostly good, but I do have one high-level comment. Would it be possible to have `path.go` and `fs_rpc.go` live in another package, say `fsrpc` as a sibling of `libkbfs`? Since that package would then depend on `libkbfs`, you'd then have to create the `fsRPC` object outside of `libkbfs`, say in `libfuse`, then plumb it through `libkbfs.Init` to `KeybaseDaemon`. That should be doable, I think.

As a side effect, that would make it easy to turn this on and off via a build flag similar to the client side, if you make it so that a `nil` `fsRPC` object means that it doesn't get used.

The main reason for this is that I think `libkbfs` should be pretty agnostic regarding the details of file paths, dir structure, etc. `libfuse` and `libdokan` (and the `kbfs` executable) do their own file-path munging, so I'd like to keep `libkbfs` agnostic.

gabriel commented:
Yup sure... its good to keep packages from knowing too much about the world... I'll move that stuff

strib commented on libkbfs/fs_rpc.go:
```diff
@@ -0,0 +1,107 @@
+	}, nil
+}
+
+// List implements keyubase1.FsInterface
```
"keyubase1" -> "keybase1"

strib commented on libkbfs/fs_rpc.go:
```diff
@@ -0,0 +1,107 @@
+func (f fsRPC) tlf(ctx context.Context, path Path) (keybase1.ListResult, error) {
+	files := []keybase1.File{}
+
+	node, de, err := path.GetNode(ctx, f.config)
```
Probably not for this PR, but note that eventually you're going to want to cache these `Node` objects somewhere, and possibly add a `Forget` RPC to remove them from the cache when the app is done with them.  Calling `GetNode` every time is needlessly expensive.  Although I guess it depends on the exact semantics we want for the mobile app, so we can cross that bridge later.  Just something to keep in mind.

strib commented on libkbfs/fs_rpc.go:
```diff
@@ -0,0 +1,107 @@
+
+		// For entryInfo: for name, entryInfo := range children
+		for name := range children {
+			files = append(files, keybase1.File{Name: name})
```
This only gives you the entry name, and not the full `/keybase/.../name` path.  Is that what you want?  The `keybase()`, and `root()` commands give the full path with a leading `/`.  I guess `favorites()` only gives the base name as well.  It should probably be consistent across all RPC types.

gabriel replied:
Will add path.Join(name) for those.

gabriel replied:
Also changed protocol so `keybase1.File{Path: "/this/is/not/name/its/path"}`

strib commented on libkbfs/fs_rpc.go:
```diff
@@ -0,0 +1,107 @@
+	}
+	files := []keybase1.File{}
+	for _, fav := range favs {
+		if (fav.Public && path.Public) || (!fav.Public && !path.Public) {
```
This can just be `if fav.Public == path.Public {`.

strib commented on libkbfs/keybase_daemon_rpc.go:
```diff
@@ -378,6 +378,7 @@ func (k *KeybaseDaemonRPC) OnConnect(ctx context.Context,
 		keybase1.IdentifyUiProtocol(daemonIdentifyUI{k.daemonLog}),
 		keybase1.NotifySessionProtocol(k),
 		keybase1.NotifyUsersProtocol(k),
+		keybase1.FsProtocol(&fsRPC{config: k.config, log: k.log}),
```
When you move `fsRPC` into a different package, you'll want to move this protocol registration there too.  You can use `libkbfs.NewSharedKeybaseConnection()` to get a shared connection, I think.

strib replied:
That probably means you'll need to add something to `kbfsfuse` and `kbfsdokan` to initialize that `fsRPC` connection, if we want it live for those binaries.  Maybe we only want it on mobile?

strib commented:
Looks good overall, just some comments.  Definitely +1 on moving `fs_rpc.go` and `path.go` into a different package.

gabriel commented:
Made the following changes:

- Moved `fs_rpc.go` and `path.go` into separate package, at top level, called just `rpc`. Is this too generic and wrong?
- Created `KeybaseDaemonFn` as option for daemon constructor (defaults to original `makeKeybaseDaemon`). Mobile will construct it's own and register additional protocols.
- Added `KeybaseDaemonRPC.AddProtocols` so the new `FsProtocol` can be added to it when mobile constructs it.

I don't know if this is better but everything is less couple-y now.

gabriel commented:
A by product of this, is someone could construct a compatible `KeybaseDaemon` and create a KBFS instance without our service :)

strib commented:
Thanks @gabriel, code looks good to me but @akalin-keybase should take a look too.

> Moved fs_rpc.go and path.go into separate package, at top level, called just rpc. Is this too generic and wrong?

It's pretty generic.  What about `fsrpc` or `rpcfs`?

strib commented:
Also:

> * Created KeybaseDaemonFn as keybase daemon constructor option for Init (defaults to original makeKeybaseDaemon). Mobile will construct it's own and register additional protocols.
> * Added KeybaseDaemonRPC.AddProtocols so the new FsProtocol can be added to it when mobile constructs it.

Will mobile actually make a different KeybaseDaemon implementation?  If it's just adding protocols, I'd favor adding a `kbdProtocols []msgpack.Protocol` to the `Init` method instead of passing in a maker function.

gabriel commented:
I thought about `fsrpc` but the fs part seemed redundant. Also I could forsee there being other RPC things that are in the kbfs "realm" but not necessarily an fs thing and wouldn't necessarily want to make yet another package for that?

I tried doing a send protocols into `libkbfs.Init` but the protocols need access to either `libkfs.Config` or the KeybaseDaemon itself. 

Mobile calls libkbfs.NewKeybaseDaemonRPC from this function:
https://github.com/keybase/client/blob/kbfsmobile-sq/go/loopback/main.go#L74-L80
(Ignore that its commented out)

I actually may need this in the future cause I kind of want to make a KeybaseDaemon that skips RPC and calls directly into the service via method calls (e.g. connection-less rpc which I guess is straight up method calling)... because on mobile this is all in a single runtime.

strib commented:
> My dream is to have KBFS on mobile not use msgpack rpc at all :)

Me too, I thought that was the plan!  What about passing in a (slice of?) callback functions that take the `KeybaseDaemonRPC` as an argument, and returns new protocols, and gets called in `KeybaseDaemonRPC.OnConnect()`?

Even better would be something that just uses `libkbfs.NewSharedKeybaseConnection` and gets its own `OnConnect` callback where it can register its own protocols, without relying on `KeybaseDaemonRPC`.

maxtaco commented:
It was my proposal to use the same technique we use to bridge with JS because I thought it would be less work, but don't have strong opinions and am happy to go with the consensus here.

gabriel commented:
Talked to @strib and we're gonna keep this `KeybaseDaemonFn` factory method with the intention of it being temporary, and in the future we'll have a separate Init method for mobile, which will be able to bypass the RPC layer and call methods directory (since client+kbfs live in the same runtime), which is out of the scope of this effort.

akalin-keybase commented:
I'm ok with `fsrpc`. `rpc` does seem too generic, and might plausibly conflict with other "real" rpc libraries.

akalin-keybase commented:
lgtm other than the `rpc` name comment.

gabriel commented:
Yep agree, I see it's ok to be unique/redundant with package names because of default import package name should be somewhat unique..

gabriel commented:
cc @chrisnojima (so he is aware of the discussion above)

chrisnojima commented:
i'm for whatever works. the original msgpack path was to keep things untouched and just to hook things up in a similar way to how it worked already (and not expose a lot of new untested code paths)
